thunder/torch/__init__.py

```diff
@@ -5331,6 +5331,230 @@ def wait(slf) -> None:
 
 
+#
+# The automatically registered torch operators
+#
+
+
+def _is_differentiable(arg: Any) -> bool:
+    from torch._subclasses.fake_tensor import FakeTensor
+
+    if isinstance(arg, (torch.Tensor, FakeTensor, TensorProxy)):
+        return dtypes.is_inexact_dtype(to_dtype(arg.dtype))
+    return False
+
+
+def _make_differentiable_wrapper(func: Callable, *args, **kwargs):
+    from thunder.core.pytree import tree_flatten, tree_unflatten
+
+    flat_args, spec = tree_flatten((args, kwargs))
+    differentiable_args_idx = tuple(i for i, a in enumerate(flat_args) if _is_differentiable(a))
+    differentiable_args = tuple(arg for i, arg in enumerate(flat_args) if i in differentiable_args_idx)
+
+    def wrapper(*diff_args):
+        diff_args_iter = iter(diff_args)
+        full_args = [next(diff_args_iter) if i in differentiable_args_idx else arg for i, arg in enumerate(flat_args)]
+        full_args, full_kwargs = tree_unflatten(full_args, spec)
+        return func(*full_args, **full_kwargs)
+
+    return wrapper, differentiable_args
+
+
+def register_default_torch_ops():
+    from thunder.torch import default_torch_ops
+
+    for m, fns in default_torch_ops.torch_auto_registered_ops.items():
+        for fn in fns:
+            register_default_torch_op(fn, meta_adaptor(fn), m)
+
+
+def register_default_torch_op(torchfn: Callable, fn_meta: Callable, m):
+    _fn = langctx(Languages.TORCH)(fn_meta)
+    sym = Symbol(
+        name=torchfn.__name__,
+        meta=_fn,
+        id=f"{m.__name__}.{torchfn.__name__}",
+    )
+    _torch_to_thunder_function_map[torchfn] = sym
+    from thunder.executors.torchex import _always_executable, ex
+
+    op = ex.register_operator(torchfn.__name__, module=m, meta=fn_meta)
+    ex.register_implementation(sym, op, checker=_always_executable)
+
+    from thunder.core.transforms import augmented_forward_impls, backward_impls
+
+    augmented_forward_impls[sym.id] = augmented_forward_adaptor(op)
+
+    def _vjp_impl(residules, *gs) -> torch.Tensor:
+        inp_args, inp_kwargs = residules
+
+        wrapped_func, diff_args = _make_differentiable_wrapper(torchfn, *inp_args, **inp_kwargs)
+        _, outs = torch.autograd.functional.vjp(wrapped_func, diff_args, v=gs)
+
+        from thunder.core.pytree import tree_unflatten, tree_flatten
+
+        flat_inp_args, inp_spec = tree_flatten(inp_args)
+        iter_out = iter(outs if isinstance(outs, Sequence) else (outs,))
+        outs = tuple(next(iter_out) if _is_differentiable(arg) else None for arg in flat_inp_args)
+        return tree_unflatten(outs, inp_spec)
+
+    bwd_op = ex.register_operator(torchfn.__name__ + "_vjp", meta=backward_adaptor(torchfn), fn=_vjp_impl)
+    ex.register_implementation(bwd_op.id, bwd_op, checker=_always_executable)
+    backward_impls[sym.id] = bwd_op
+
+
+def _get_fake_arg(inp: Any, fake_mode):
+    from thunder.core.devices import to_torch_device
+    from thunder.core.proxies import TupleProxy, ListProxy, DictProxy
```

Comment:
Do you mean move it to the top of this file? I thought only this func uses them and put them here https://github.com/Lightning-AI/lightning-thunder/blob/3c315b4ca14ada9159c4350f9fb058321a29f7b9/thunder/torch/__init__.py#L30-L36

Reply:
As per pep8 guidelines - import should be on top - I think one of the exception is when they would lead to cyclical dependency (or maybe more) but since we already import `thunder.core` in this file, I think it would be better to move them at the beginning of the file.

https://peps.python.org/pep-0008/#imports

Another discussion on stackoverflow - https://stackoverflow.com/questions/128478/should-import-statements-always-be-at-the-top-of-a-module